### ISSUE TITLE
NO use of GPU?How about the trainning time

### ISSUE THREAD

**Issue by hejunqing** [mkroutikov/tf-lstm-char-cnn#3]
Hi,I am a freshman in tf.  Having a glance on the Readme and your code,I can't find any thing like CUDA and " if device_name == "gpu": ". So the code is running without GPU? How about the training time for 25 epochs?

**Comment by hejunqing** [comment]
According to  tensorflow document, the program runs with GPU if the  there is GPU implement.It took more than 20 hours to train the model on a K20 with GPU, but most of the time the load is 0%. It is also strange that the first epoch secs/batch = 0.8538s, the last epoch secs/batch = 4.0487s.The time for each batch gets longer from the first to last epoch.Why?

**Comment by mkroutikov** [comment]
Yes, I see the same.

1. Torch7 code (the original) is 20% faster and per-batch time does not grow
2. GPU does not help. It is trained faster on 32 cpu machine than on GPU
3. training time around 20 hours is correct - I see same here
4. unexplained slowdown as system learns

about (4): 
since code re-starts training from the last saved checkpoint, it is possible to interrupt (Ctrl+C) training, and then re-start. Could you please confirm that after re-starting it is still as slow (or its back to fast time per batch)?

**Comment by hejunqing** [comment]
I change some of your code and receive a great improvement on speed! Now the training time is secs/batch =0.4021s and <550s per epoch on K20m.And it won't get slower after epochs!It takes about 4 hours to train now.OH yeah~

Here are changes:

 - Substitute all the For loops with matrix operation and change the conv style(the same result but more efficient and easy-to-understand). This cuts the time by around 0.04 secs/batch.
 - Separate the graph and session in train.py. Put all the define things into the graph, and only operate in session.The define things in loops in Graph will add nodes to it, which is reason for low load on GPU.
 - Remove the clear_embedding_function and do the matrix assign thing in the inference_graph.This cuts the time from 0.8 to 0.4 secs/batch

I am working to make the codes shorter. If anybody needs my code, I will put on Github later.
@mkroutikov Thanks for quick reply,your are so nice.

**Comment by mkroutikov** [comment]
This is super cool. Please, do push your changes to the GitHub, and create a pull request!

**Comment by mkroutikov** [comment]
Nicole, if you do not have time for code cleanup or pull request, please just share your code --- I'll analyze and merge it in.

So, the slowdown was caused by the fact that I was doing "session.run(tf.assign(...))" in the learning loop, which kept adding assign operation to the graph... Thanks for this research.

I'd like to see your conv ops speedups as well.

The reason for the existence of clear_embedding_padding() function is to make graph computations identical to Torch7 code. There, embedding is initialized such that the embedding vector of the padding symbol is zero. And Torch7 has special handling of learning updates to keep it zero during the training.

Torch7 comments seem to indicate that this makes embedding op faster. Strictly speaking, there is nothing wrong in removing this operation and letting model learn the "best" embedding for the padding symbol in TF. I just was comparing tensor values from Torch7 and TF at every processing step, to make sure that TF code is identical. Now that I know that it *is* identical, we can tweak and simplify it.

cc: @hejunqing

**Comment by hejunqing** [comment]
Yes,the clear_embedding _padding() is just to make the blank embedding keep zero.
Result goes worse  a bit if just remove it. So I put the zero assigning of the word_embedding in the inference_graph right before embedding lookup.
I am not so skilled at lua and Torch, where is zero pad handling in lua code ? I will find if there is similar handling in TF. It will run faster is we can stop it at gradient learning.

**Comment by mkroutikov** [comment]
1. Please, pull - I removed the offending function and zero embedding slot using OP instead
2. Maybe a better way to do this is to use a custom initializer and gradient mask. But I am not an expert, so will go for the current solution
3. Yoon Kim's Torch code does zero out padding vector here:
   https://github.com/yoonkim/lstm-char-cnn/blob/master/main.lua#L371
   and after each gradient update here:
   https://github.com/yoonkim/lstm-char-cnn/blob/master/main.lua#L380

   Note that Torch's LookupTable natively supports not updating padding, but Kim does not use it:
        https://github.com/torch/nn/blob/master/lib/THNN/generic/LookupTable.c#L112
        (note that gradient propagation is skipped for padding index)

**Comment by hejunqing** [comment]
I push my code eventually~I am new to github and don't know how to commit to your repo.So I made my repo.
 [myrepo](https://github.com/hejunqing/tf-char-cnn-lstm)
I have watched your new code, nice job! @mkroutikov

**Comment by mkroutikov** [comment]
Thank you, Nicole, for sharing your code. Here is the summary of the significant changes:

1. zero out RNN state at each epoch. The original Yoon Kim's code does not do this (unless I misunderstood the code), see https://github.com/yoonkim/lstm-char-cnn/blob/master/main.lua#L306 . From the other hand, that looks like a reasonable change, because it does not seem right to keep RNN state after word  sequence wraps up to the start. For now I am not changing this as the first goal was to port Yoon Kim's code as-is.

2. Fix performance bug. Already incorporated. Thanks for finding and reporting it.

3. You do not unroll RNN tensors into a list, but operate on the packed tensor. This may give some performance improvements. I need to research this.

By the way, your name is now in the list of contributors in the README!

**Comment by mkroutikov** [comment]
I ran some experiments with my code and your `train_gpu.py` on AWS machines.

I do not see any significant difference in speed. Your code seem to be 0.5% faster on GPU and 2% slower on CPU, but these results are not statistically significant.

See new *Training time* section in the README.

Here are the comparative timings (number of seconds to complete one epoch):

|  System  |  `c4.8xlarge` | `g2.2xlarge` | `p2.xlarge` |
|-------------| ------------------|------------------| ---------------|
| Mike's    | 1404 sec        | 3779 sec     | 428 sec     |
| Nicole's | 1426 sec       | 3758 sec     | 425 sec      |
| diff         |  +1.6%          |   -0.6%        |  -0.7%        |
